pool.go

```diff
@@ -0,0 +1,244 @@
+//Package ggpool provides functionality of simple generic objects pool
+package ggpool
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"sync"
+	"time"
+)
+
+//Pool is a pool of generic objects
+type Pool struct {
+	config                Config
+	itemReleasedCh        chan bool
+	hasPendingNewItemCh   chan bool
+	createItemLastErrorCh chan error
+	ctx                   context.Context
+	cancel                context.CancelFunc
+
+	sync.RWMutex
+	itemCollection *collection
+}
+
+//NewPool returns a new Pool instanse
+func NewPool(ctx context.Context, config Config) (*Pool, error) {
+	var p *Pool
+
+	if err := config.validate(); err != nil {
+		return p, err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
Mb it's better to control context outside of the pool:
- Remove pool.Cancel()
- Provide context.WithCancel() to NewPool func